### PR TITLE
764 feature html export

### LINKED ISSUE #764: Export in HTML format
Add default format for export (html)

### PR DESCRIPTION
**Pull request by BobTheSuperman**

### CODE REVIEW

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+using System.Text;
+
+namespace CharlieBackend.Business.Services.FileServices.ExportFileServices.Html
+{
+    class ClassbookExportHTML : BaseFileExportHTML
+    {
+        public ClassbookExportHTML()
+        {
+
+        }
```
Comment (AccorDDD) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs]: Why is this constructor here?

**Review by AccorDDD — APPROVED** [review]

**Review by leshikevich — APPROVED** [review]

**Review by Ssashaa — APPROVED** [review]

**Review by Soctarian — APPROVED** [review]

File: CharlieBackend.Business/Helpers/HtmlGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+                tableBody.Append("</tr>");
+                counter++;
+            }
+            tableBody.Append("</tbody>");
+
+            table.Append(tableHead);
+            table.Append(tableBody);
+            table.Append("</table>");
+
+
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Business/Helpers/HtmlGenerator.cs]: please remove space

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/ExportServiceHtml.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
+﻿using CharlieBackend.Business.Services.FileServices.ExportFileServices.Html;
+using CharlieBackend.Core.DTO.Dashboard;
+using CharlieBackend.Core.DTO.Export;
+using CharlieBackend.Core.Models.ResultModel;
+using CharlieBackend.Data.Repositories.Impl.Interfaces;
+using System.Threading.Tasks;
+
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/ExportServiceHtml.cs:8]: please remove space

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/ExportServiceHtml.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                {
+                    ByteArray = await fileExporter.GetByteArrayAsync(),
+                    ContentType = fileExporter.GetContentType(),
+                    Filename = fileExporter.GetFileName()
+                });
+            }
+            else
+            {
+                return Result<FileDto>.GetError(ErrorCode.NotFound,
+                        "Group not found");
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/ExportServiceHtml.cs]: please extract to response messages

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+                StringBuilder html = HtmlGenerator.GenerateHtml(GetFileHeader(data), table);
+
+                byte[] byteLine = ConvertLineToArray(html.ToString());
+
+                _memoryStream.Write(byteLine);
+            }
+        }
+
+        private void TryToFillMarks(StudentsClassbookResultDto data)
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs]: please rename to fill marks

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        {
+            if (data.StudentsMarks != null && data.StudentsMarks.Any())
+            {
+                var dateData = data.StudentsMarks.GroupBy(x => x.LessonId);
+
+                string[] headers = new string[dateData.Count() + 2];
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs:75]: please leave a comment

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+                var students = data.StudentsMarks.Select(x => x.Student).Distinct().OrderByDescending(s => s);
+
+                string[][] rows = new string[students.Count()][];
+
+                for (int i = 0; i < rows.Length; i++)
+                {
+                    rows[i] = new string[headers.Length];
+                    rows[i][0] = students.ElementAt(i);
+
+                    List<int> listOfMarks = new List<int>();
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs]: please rename

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+                            .Where(x => x.Student == rows[i][0] && x.LessonDate?
+                            .ToString("dd.MM.yyyy") == headers[j + 1]);
+
+                        var studentMark = marks.Any() ? marks.First().StudentMark : null;
+
+                        rows[i][j + 1] = studentMark == null ? " " : studentMark == 0 ? "-" : $"{studentMark}";
+
+                        if(rows[i][j + 1] != " "  && rows[i][j + 1] != "-")
+                        {
+                            listOfMarks.Add(Convert.ToInt32(rows[i][j + 1]));
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs]: please rework according to conversation

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+                finishDate = data.StudentsPresences.Last().LessonDate?.ToString("dd.MM.yyyy");
+            }
+            else if(data.StudentsMarks != null && data.StudentsMarks.Any())
+            {
+                group = data.StudentsMarks.First().StudentGroup;
+                startDate = data.StudentsMarks.First().LessonDate?.ToString("dd.MM.yyyy");
+                finishDate = data.StudentsMarks.Last().LessonDate?.ToString("dd.MM.yyyy");
+            }
+
+            return $"Group: {group} {startDate}&nbsp;-&nbsp;{finishDate}";
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs]: please move to constants

**Review by ArsenMaistryshyn — CHANGES_REQUESTED** [review]

File: CharlieBackend.Business/Helpers/HtmlGenerator.cs (hunk 2 above)
Comment (BobTheSuperman) [reply CharlieBackend.Business/Helpers/HtmlGenerator.cs]: Done!

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/ExportServiceHtml.cs (hunk 3 above)
Comment (BobTheSuperman) [reply CharlieBackend.Business/Services/FileServices/ExportFileServices/ExportServiceHtml.cs:8]: Done!

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs (hunk 5 above)
Comment (BobTheSuperman) [reply CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs]: Done!

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs (hunk 7 above)
Comment (BobTheSuperman) [reply CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs]: Done

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs (hunk 8 above)
Comment (BobTheSuperman) [reply CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs]: Done!

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs (hunk 1 above)
Comment (BobTheSuperman) [reply CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs]: Removed it, ty.

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/ExportServiceHtml.cs (hunk 4 above)
Comment (BobTheSuperman) [reply CharlieBackend.Business/Services/FileServices/ExportFileServices/ExportServiceHtml.cs]: Done

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs (hunk 9 above)
Comment (BobTheSuperman) [reply CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs]: Done!

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs (hunk 6 above)
Comment (BobTheSuperman) [reply CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookExportHTML.cs:75]: Done!

**Review by ArsenMaistryshyn — APPROVED** [review]

**Pull request merged**